src/js/mqtt.js

```diff
@@ -0,0 +1,308 @@
+/* Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var net = require('net');
+var util = require('util');
+var EventEmitter = require('events').EventEmitter;
+
+util.inherits(MQTTClient, EventEmitter);
+
+PacketTypeEnum = {
+  PUBACK : 4,
+  PUBREC : 5,
+  PUBREL : 6,
+  PUBCOMP : 7
+};
+
+function MQTTClient(options) {
+  if (!(this instanceof MQTTClient)) {
+    return new MQTTClient(options);
+  }
+
+  EventEmitter.call(this);
+
+  this._protocol = options.protocol || 'net';
+
+  this._clientOptions = Object.create(options, {
+    host: { value: options.host || '127.0.0.1'},
+    port: { value: options.port || 8883 },
+    qos: { value: options.qos || 0 },
+    keepalive: { value: options.keepalive || 60 },
+  });
+
+  this._socket = new net.Socket();
```

Comment:
Is it a simple `net.Socket` even if tls is specified? I would remove the _protocol member and would simply accept a duplex stream as options.socket. If it is not specified, a new `net.Socket` is created.